apps/accelerate/chatllama/chatllama/llama_model.py

```diff
@@ -59,50 +414,109 @@ class Transformer(nn.Module):
+        self.cache_v = [torch.zeros(dim) for _ in range(self.n_layers)]
+
+        if params.use_fairscale:
+            self.tok_embeddings = ParallelEmbedding(
+                params.vocab_size, params.dim, init_method=lambda x: x
+            )
+        else:
+            self.tok_embeddings = nn.Embedding(params.vocab_size, params.dim)
+
+        if params.forze_embeddings:
+            for param in self.tok_embeddings.parameters():
+                param.requires_grad = False
 
         self.layers = torch.nn.ModuleList()
         for layer_id in range(params.n_layers):
             self.layers.append(TransformerBlock(layer_id, params))
 
         self.norm = RMSNorm(params.dim, eps=params.norm_eps)
-        self.output = ColumnParallelLinear(
-            params.dim, params.vocab_size, bias=False, init_method=lambda x: x
-        )
+        if params.use_fairscale:
+            self.output = ColumnParallelLinear(
+                params.dim,
+                params.vocab_size,
+                bias=False,
+                init_method=lambda x: x,
+            )
+        else:
+            self.output = nn.Linear(params.dim, params.vocab_size, bias=False)
 
+        # TODO: How too modify this for training?
         self.freqs_cis = precompute_freqs_cis(
             self.params.dim // self.params.n_heads, self.params.max_seq_len * 2
         )
 
-    def forward(self, tokens: torch.Tensor, attention_mask: torch.Tensor):
-        start_pos = int(torch.argmax(attention_mask.detach(), dim=-1).item())
-        logits = self._forward(tokens, start_pos)
+    def forward(
+        self, tokens: torch.Tensor, mask: torch.Tensor
+    ) -> torch.Tensor:
+        mask = mask.detach()
+        logits = self._forward(tokens, mask)
         return logits
 
-    def _forward(self, tokens: torch.Tensor, start_pos: int):
+    def _forward(
+        self, tokens: torch.Tensor, mask: torch.Tensor
+    ) -> torch.Tensor:
         _bsz, seqlen = tokens.shape
         h = self.tok_embeddings(tokens)
         self.freqs_cis = self.freqs_cis.to(h.device)
+        start_pos = 1  # TEMPORARY
         freqs_cis = self.freqs_cis[start_pos : start_pos + seqlen]  # noqa E203
+        attention_mask = torch.zeros(
+            (_bsz, seqlen, seqlen), device=mask.device
+        )
+        # if the mask is a boolean tensor, convert it to int
+        if mask.dtype == torch.bool:
+            mask = mask.long()
+        for i in range(tokens.shape[0]):
+            start_pos = torch.argmax(mask[i])
+            for j in range(tokens.shape[1]):
+                if mask[i, j] != 0:
+                    attention_mask[i, j, start_pos : j + 1] = 1  # noqa E203
+        attention_mask = torch.where(
+            attention_mask == 0, float("-inf"), attention_mask
+        )
+        attention_mask = (
+            torch.where(attention_mask == 1, float(0.0), attention_mask)
+            .long()
+            .detach()
+        )
```

Comment:
```suggestion
        # mask has size (bsz, seqlen). It should be transformed in (bsz, seqlen, seqlen)
        # if the mask is a boolean tensor, convert it to int
        if mask.dtype == torch.bool:
            mask = mask.long()
        attention_mask = mask[:, None, None, :].expand(_bsz, 1, seqlen, seqlen)
        attention_mask = 1 - attention_mask
        attention_mask = torch.where(
            attention_mask == 1, float("-inf"), attention_mask
        ).long().detach()
```

the code above should be roughly equivalent to what you wrote. But isn't this missing the `triu` part? :)

Reply:
This is not working as it is. 
if i have a vect [0, 0, 0, 1, 1, 1]
I should produce [
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 0, 1, 1, 0],
    [0, 0, 0, 1, 1, 1]
]
The output of this code will be:
I should produce [
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1]
]
i can find a easy way to do this with the triu without overcomplicating this.

Reply:
if you take the `torch.tril` matrix of what I wrote you get the correct one :)